### PR TITLE
Backend topics changing

### PR DESCRIPTION
**Pull request by yazebochan**

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java [hunk 1]
```diff
@@ -62,6 +63,14 @@
         return kafkaService.createTopic(cluster, topicFormData);
     }
 
+    public Mono<ResponseEntity<Topic>> updateTopic(String clusterName, String topicName, Mono<TopicFormData> topicFormData) {
+        KafkaCluster cluster = clustersStorage.getClusterByName(clusterName);
+        if (cluster == null) return null;
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java:68]: This is bad idea

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java [hunk 2]
```diff
@@ -62,6 +63,14 @@
...
+        return ClusterUtil.toMono(cluster.getAdminClient().describeCluster().controller())
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java:69]: We talked to move this logic outside

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java [hunk 3]
```diff
@@ -72,6 +77,27 @@ public void loadClusterMetrics(KafkaCluster kafkaCluster) {
         );
     }
 
+    public Mono<Topic> updateTopic(KafkaCluster cluster, String topicName, TopicFormData topicFormData, Integer id) {
+        ConfigResource topicCR = new ConfigResource(ConfigResource.Type.TOPIC, topicName);
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java:81]: Please refactor it

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java [hunk 4]
```diff
@@ -72,6 +77,27 @@ public void loadClusterMetrics(KafkaCluster kafkaCluster) {
...
+        List<ConfigResource> brokerCR = Collections.singletonList(new ConfigResource(ConfigResource.Type.BROKER, id.toString()));
+        return ClusterUtil.toMono(cluster.getAdminClient().describeConfigs(brokerCR).all())
+                .flatMap(c -> {
+                    if (oldClusterVersion(c, cluster)) {
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java:85]: we could cache cluster version on adminclient cache level.

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/zookeeper/ZooKeeperConstants.java [hunk 5]
```diff
@@ -7,4 +7,7 @@ private ZooKeeperConstants() {}
     public static int ONLINE = 1;
     public static int OFFLINE = 0;
 
+    public static int CONNECTION_TIMEOUT_MS = 1000;
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/zookeeper/ZooKeeperConstants.java:10]: move to config

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 6]
```diff
@@ -141,6 +141,34 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/TopicDetails'
+    patch:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:144]: Why not PUT?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java (hunk 1 above)
Comment (yazebochan) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java]: changed to return Mono.error

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java (hunk 2 above)
Comment (yazebochan) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/ClusterService.java]: moved to controller level

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java (hunk 4 above)
Comment (yazebochan) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java]: cluster supported commands enum added to KafkaCluster object

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/zookeeper/ZooKeeperConstants.java (hunk 5 above)
Comment (yazebochan) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/zookeeper/ZooKeeperConstants.java]: moved

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 6 above)
Comment (yazebochan) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml]: changed to put

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/util/ClusterUtil.java [hunk 7]
```diff
@@ -31,4 +39,22 @@ public static ConsumerGroup convertToConsumerGroup(ConsumerGroupDescription c, K
         consumerGroup.setNumTopics(topics.size());
         return consumerGroup;
     }
+
+    public static void setSupportedCommands(KafkaCluster cluster, Map<ConfigResource, Config> configs) {
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/util/ClusterUtil.java:43]: this method should provide immutable list of supported features

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java [hunk 8]
```diff
@@ -72,41 +76,31 @@ public void loadClusterMetrics(KafkaCluster kafkaCluster) {
...
-        } catch (Exception e) {
-            log.error(e);
-            kafkaCluster.setLastKafkaException(e);
-
-            return false;
-        }
+    public Mono<Topic> updateTopic(KafkaCluster cluster, String topicName, TopicFormData topicFormData, Integer id) {
+        ConfigResource topicCR = new ConfigResource(ConfigResource.Type.TOPIC, topicName);
+        List<ConfigResource> brokerCR = Collections.singletonList(new ConfigResource(ConfigResource.Type.BROKER, id.toString()));
+        return ClusterUtil.toMono(cluster.getAdminClient().describeConfigs(brokerCR).all())
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java:82]: You can do it only once on adminclient init

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java [hunk 9]
```diff
@@ -72,41 +76,31 @@ public void loadClusterMetrics(KafkaCluster kafkaCluster) {
...
+                .flatMap(c -> {
+                    if (cluster.getSupportedCommands().isEmpty()) {
+                        ClusterUtil.setSupportedCommands(cluster, c);
+                    }
+                    if (cluster.getSupportedCommands().contains(SupportedCommands.INCREMENTAL_ALTER_CONFIGS)) {
+                        List<AlterConfigOp> listOp = topicFormData.getConfigs().entrySet().stream()
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java:88]: I'll separate these methods for clear code

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java [hunk 10]
```diff
@@ -267,4 +261,40 @@ private void createTopic(AdminClient adminClient, NewTopic newTopic) {
...
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            kafkaCluster.setLastKafkaException(e);
+
+            return false;
+        }
+    }
+
+    @SneakyThrows
+    private String getClusterId(KafkaCluster kafkaCluster) {
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java:284]: Blocking method?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java (hunk 9 above)
Comment (yazebochan) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/kafka/KafkaService.java]: separated

**Pull request merged**